_posts/2024-06-14-granite-rag.adoc

```diff
@@ -0,0 +1,286 @@
+
+We also need a specific dependency to use the BGE-small embedding model:
+
+[source,xml]
+----
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-embeddings-bge-small-en-q</artifactId>
+    <version>0.31.0</version>
+</dependency>
+----
+
+=== Configuration
+
+We need a bit of configuration to ensure our application uses Granite and sets up the Infinispan database correctly:
+
+[source, properties]
+----
+# Configure the Infinispan vectors:
+quarkus.langchain4j.infinispan.dimension=384
+
+# Configure the OpenAI service to use instruct lab:
+quarkus.langchain4j.openai.base-url=http://localhost:8000/v1
+quarkus.langchain4j.openai.timeout=60s
+
+quarkus.langchain4j.embedding-model.provider=dev.langchain4j.model.embedding.BgeSmallEnQuantizedEmbeddingModel
+----
+
+The first line configures the dimension of the vectors stored in Infinispan, which depends on the embedding model (BGE-small in our case).
+The second and third lines configure the OpenAI service to use Instructlab.
+You can replace the base URL with the one for Podman AI Studio if you prefer.
+
+The last line sets the default embedding model to BGE-small.
+
+=== The Ingestor
+
+With the configuration in place, let's implement the https://github.com/cescoffier/quarkus-granite-rag-demo/blob/main/src/main/java/me/escoffier/granite/rag/Ingestion.java[ingestor].
+The ingestor reads documents from the documents directory, splits them into text segments, computes their vector representations, and stores them in Infinispan:
+
+[source,java]
+----
+@Singleton
+@Startup
+public class Ingestion {
+
+    public Ingestion(EmbeddingStore<TextSegment> store, EmbeddingModel embedding) {
+
+        EmbeddingStoreIngestor ingestor = EmbeddingStoreIngestor.builder()
+                .embeddingStore(store)
+                .embeddingModel(embedding)
+                .documentSplitter(recursive(1024, 0))
+                .build();
+
+        File file = new File("documents");
+        List<Document> documents = FileSystemDocumentLoader.loadDocuments(file.toPath());
+        Log.info("Ingesting " + documents.size() + " documents");
+
+        ingestor.ingest(documents);
+
+        Log.info("Document ingested");
+    }
+
+}
+----
+
+The `@Startup` annotation ensures that the ingestion process starts when the application launches.
+The `Ingestion` class uses an `EmbeddingStore<TextSegment>` (Infinispan) and an `EmbeddingModel` (BGE-small).
+
+We use a simple document splitter (`recursive(1024, 0)`) to split the documents into text segments.
+More advanced techniques may be used to improve the accuracy of the RAG model.
+
+=== The retriever
+
+Next, let's implement the https://github.com/cescoffier/quarkus-granite-rag-demo/blob/main/src/main/java/me/escoffier/granite/rag/Retriever.java[retriever].
+The retriever finds the most relevant text segments in Infinispan based on the user query:
+
+[source,java]
+----
+@Singleton
```

Comment:
From our experience, it's better to provide the imports.

Reply:
I've linked the full class.